llvm/lib/Transforms/Utils/SplitModule.cpp

```diff
@@ -268,6 +268,38 @@ void llvm::SplitModule(
   ClusterIDMapType ClusterIDMap;
   findPartitions(M, ClusterIDMap, N);
 
+  // Find empty modules and functions not mapped to modules in ClusterIDMap.
+  // Map these functions to the empty modules so that they skip being
+  // distributed by isInPartition() based on function name hashes below.
+  // This provides better uniformity of distribution of functions to modules
+  // in some cases - for example when the number of functions equals to N.
+  if (TryToAvoidEmptyModules) {
+    DenseSet<unsigned> NonEmptyModules;
+    SmallVector<const GlobalValue *> UnmappedFunctions;
+    for (const auto &F : M.functions()) {
+      if (F.isDeclaration() ||
+          F.getLinkage() != GlobalValue::LinkageTypes::ExternalLinkage)
+        continue;
+      auto It = ClusterIDMap.find(&F);
+      if (It == ClusterIDMap.end())
+        UnmappedFunctions.push_back(&F);
+      else
+        NonEmptyModules.insert(It->second);
+    }
+    SmallVector<unsigned> EmptyModules;
+    for (unsigned I = 0; I < N; ++I) {
+      if (!NonEmptyModules.contains(I))
+        EmptyModules.push_back(I);
+    }
+    auto NextEmptyModuleIt = EmptyModules.begin();
+    for (const auto F : UnmappedFunctions) {
+      if (NextEmptyModuleIt == EmptyModules.end())
+        break;
+      ClusterIDMap.insert({F, *NextEmptyModuleIt});
+      ++NextEmptyModuleIt;
+    }
```

Comment:
I don't quite get this loop, seems like you won't distribute all the unmapped functions, but only at most a number equal to the empty modules? All the others will be distributed by isInPartition() if I read this loop correctly.

Why don't you just circle back to `EmptyModules.begin()` instead of breaking out of the loop?

Also should we skip the loop over UnmappedFunctions if `EmptyModules.empty()` ?

Reply:
> I don't quite get this loop, seems like you won't distribute all the unmapped functions, but only at most a number equal to the empty modules? All the others will be distributed by isInPartition() if I read this loop correctly.

Correct. Our main use case for this is to request the number of modules equal to the number of functions - then (provided that the functions are independent) the algorithm generates exactly 1 function per module. The behavior corresponds to the name of the flag - TryToAvoidEmptyModules.

> Why don't you just circle back to EmptyModules.begin() instead of breaking out of the loop?

Can do that too - this would become a different algorithm, completely overriding the hash-based distribution. For our goal (1 function per module) this would work equally. Do you think we should change it this way?

> Also should we skip the loop over UnmappedFunctions if EmptyModules.empty() ?

Indeed. Fixed.

Reply:
> Our main use case for this is to request the number of modules equal to the number of functions

I see: that is quite specific, it also makes the algorithm (and flags) only really suited for this case it seems. The name  "TryToAvoidEmptyModules" is explicit, but one may wonder "why would anyone use this algorithm as is?" without knowing the implicit assumption about "the number of modules equal to the number of functions".

Reply:
>"why would anyone use this algorithm as is?"

To get less useless empty modules.

> only really suited for this case it seems

It's best suited for this use case but works fine in the other cases too and does the same - reduces the number of empty modules as compared to what hash-based distribution does.

Reply:
> To get less useless empty modules.
>  reduces the number of empty modules as compared to what hash-based distribution does.

Right, I know it'll "work", but "to get less useless empty modules." isn't an obvious optimization metrics to me! It just seems like an odd incomplete solution to a load-balancing problem.

Reply:
As suggested I extended the new algorithm to do round-robin of all unmapped functions completely overriding the name-hash-based distribution. This works the same way for our main use case (number of modules equal to the number of functions) and may be more useful in other use cases. In the updated test this produces 2+2 function modules instead of the default 1+3.